### PR TITLE
AP_Logger: Add DataFlash log AHRS ground speed & vertical speed on AHR2

### PR DESCRIPTION
**Pull request by Jaaaky**
Log AHRS ground speed aka ahrs.groundspeed() and  vertical speed aka -ahrs.get_velocity_NED(vel).z to DataFlash logs.
Very useful to debug non-GPS navigation or dead-reckoning.

### CODE REVIEW

**Comment by OXINARF** [comment]
@Jaaaky Have you tested this? It is a bit suspicious that all SITL test jobs are failing.

**Comment by Jaaaky** [comment]
> @Jaaaky Have you tested this? It is a bit suspicious that all SITL test jobs are failing.

Yes I've tested this and it is working on SITL like charm.

**Comment by Jaaaky** [comment]
@OXINARF Should be fixed now! I don't know why it was working fine for me!
It's was sitl specific issue.

**Comment by Jaaaky** [comment]
@OXINARF All checks have passed! So waiting @peterbarker for review

**Comment by peterbarker** [comment]
@Jaaaky Ah!  Right.  So I've wanted this for a long time.

We have a `POS` message which gives you the vehicle's canonical position.

I can't see any reason we shouldn't have a `VEL` message which gives you the vehicle's canonical vehicle velocities, NED.

I think that's a better idea than throwing it into AHR2; that's supposed to be just for secondary (typically DCM) solution, not for whatever the ahrs tells us is the groundspeed.

**Comment by Jaaaky** [comment]
So add AHRS VEL MSG type with velocities NED and groundspeeed (I know it could be calced used NED but for users it's not trivial)?
I would then also add wind estimate then too. To view high-level wind speed/direction.

**Comment by Jaaaky** [comment]
One question. Is velocity.z assumed perpendicular to vehicle x plane or to earth? If plane is going up/down at 45 pitch degrees for example these is difference then between named "z" and VS.

**Comment by peterbarker** [comment]
> So add AHRS VEL MSG type with velocities NED and groundspeeed (I know it could be calced used NED but for users it's not trivial)?

Users are expected to use tools to extract data from these logs.  We try not to add derived values into these logs.

And just like POS - just `Write_VEL`, `LOG_VEL_MSG` and `log_VEL` (structure).  

> I would then also add wind estimate then too. To view high-level wind speed/direction.

We already do have the WIND message for that...

**Comment by peterbarker** [comment]
velocities should be NED (@WickedShell @rmackay9 @tridge for confirmation on that, however :-)
)

If the vehicle is pitched down at 45 degrees but fighting a (rather significant) thermal and maintaining the same absolute altitude then D should be 0, despite it being non-zero in body-frame (assuming the vehicle is moving forward ;-) )

**Comment by Jaaaky** [comment]
@peterbarker Check now. I've removed the ahr2 commit and added log_VEL.

File: libraries/AP_Logger/LogFile.cpp [hunk 1]
```diff
@@ -500,6 +500,21 @@ void AP_Logger::Write_POS(AP_AHRS &ahrs)
...
 }
 
+// Write a VEL packet
+void AP_Logger::Write_VEL(AP_AHRS &ahrs)
+{
+    Vector3f vel;
+    struct log_VEL pkt = {
+        LOG_PACKET_HEADER_INIT(LOG_VEL_MSG),
+        time_us : AP_HAL::micros64(),
+        vx      : ahrs.get_velocity_NED(vel) ? vel.x : quiet_nanf(),
```
Comment (peterbarker) [inline libraries/AP_Logger/LogFile.cpp]: Calling this method three times seems a little much :-)
Probably want a `bool have_vel = ahrs.get_velocity_NED(vel);` outside the struct definition.

File: libraries/AP_Logger/LogStructure.h [hunk 2]
```diff
@@ -417,6 +417,12 @@ struct PACKED log_POS {
     float rel_origin_alt;
 };
 
+struct PACKED log_VEL {
+    LOG_PACKET_HEADER;
+    uint64_t time_us;
+    float vx, vy, vz, gs;
```
[the remark this answers is not in the document]
Comment (Jaaaky) [reply libraries/AP_Logger/LogStructure.h]: OK, But why? I can see other code examples does the same on the same header file.

File: libraries/AP_Logger/LogFile.cpp (hunk 1 above)
Comment (Jaaaky) [reply libraries/AP_Logger/LogFile.cpp]: Yeah, so nice. I thought it was not good but didn't get better idea to do it.

File: APMrover2/Log.cpp [hunk 3]
```diff
@@ -17,6 +17,7 @@ void Rover::Log_Write_Attitude()
     logger.Write_AHRS2(ahrs);
 #endif
     logger.Write_POS(ahrs);
+    logger.Write_VEL(ahrs);
```
[the remark this answers is not in the document]
Comment (Jaaaky) [reply APMrover2/Log.cpp]: It's same for     
```
void Write_AHRS2(AP_AHRS &ahrs);
void Write_POS(AP_AHRS &ahrs);
void Write_VEL(AP_AHRS &ahrs);
void Write_EKF(AP_AHRS_NavEKF &ahrs);
```

Should it be fixed for all of them?

[the remark this answers is not in the document]
Comment (Jaaaky) [reply APMrover2/Log.cpp]: All changes are done except this, waiting your response here.

**Comment by peterbarker** [comment]
On Fri, 21 Jun 2019, Jaaaky wrote:

> Should it be fixed for all of them?

Yep!  Not on this PR, 'though :-)

**Comment by peterbarker** [comment]
On Fri, 21 Jun 2019, Jaaaky wrote:

> +    uint64_t time_us;
> +    float vx, vy, vz, gs;
> 
> OK, But why? I can see other code examples does the same on the same header
> file.

All good, then.

It's a stylistic thing we don't see of much in ArduPilot, but it's not a 
biggy.

File: APMrover2/Log.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jaaaky) [reply APMrover2/Log.cpp]: But why would it be better not to get just the referrer instead of calling singleton?

**Comment by Jaaaky** [comment]
> On Fri, 21 Jun 2019, Jaaaky wrote: Should it be fixed for all of them?
> Yep! Not on this PR, 'though :-)

@peterbarker Done for Write_VEL

Any changes still requested?

**Comment by Jaaaky** [comment]
What prevents this from being merged?

**Comment by peterbarker** [comment]
@Jaaaky 

The notes from the devcall read like this:

```
UTC2359 - https://github.com/ArduPilot/ardupilot/pull/11604
Extra CPU + bandwidth
Can we justify it?
Pos was added so we could plot canonical position on map
EKF doesn’t log lat/lon
Adding a field to add which estimator we are using
Need to know how much bandwidth this uses
Need to know how much CPU this ues
Should we combine POS and VEL?
```

(we're trying to get better about making sure these issues get this information into them during the devcall, rather than being prompted for it 8 weeks later!)

**Comment by amilcarlucas** [comment]
Needs a big rebase. Does it still make sense?

**Comment by Jaaaky** [comment]
> Needs a big rebase. Does it still make sense?

I think it does, I cannot explain how useful it was for me debugging many issues and performance.

I'm ready for a rebase if it'd get merged..

**Comment by amilcarlucas** [comment]
@Jaaaky can you answer some of the questions raised on the DevCall meeting notes posted above by Peter?

**Comment by davidbuzz** [comment]
@peterbarker  - if we are concerned about its cpu-usage, could it be enabled on demand , and default-to-off?

**Pull request closed without merging**